Readme.md
# selfish

[![Build Status](https://secure.travis-ci.org/Gozala/selfish.png)](http://travis-ci.org/Gozala/selfish)

Class-free, pure prototypal multiple inheritance that lets you write expressive,
well-structured code.

## Install

### server-side

```bash
npm install selfish
```

### client-side

```bash
bower install selfish
```

## Require

### server-side

```js
var Base = require('!raw.github.com/Gozala/selfish/v1.0.0/selfish').Base
```

### client-side RequireJS

```js
define(['path/to/selfish'], function(selfish) {
   var Base = selfish.Base;
})
```

## Examples

### Basics

```js
// Instead of creating classes, you create prototype objects. Let's look
// at this simple example first:
var Dog = Base.extend({
  bark: function() {
    return 'Ruff! Ruff!'
  }
})

// Forget about classes, javascript is a prototypal language:
typeof Dog                // object

// Use the new operator to create an instance:
var dog = new Dog()
dog.bark()                // 'Ruff! Ruff!'

// Alternatively you can use the legacy new() function but keep in mind that
// the new operator is faster in most browsers
var dog = Dog.new()
dog.bark()                // 'Ruff! Ruff!'

// Forget about special `instanceof` operator, use JS native method instead:
Dog.prototype.isPrototypeOf(dog)    // true

// Objects inherit from objects, what could be more object oriented than
// that ?
var Pet = Dog.extend({
  initialize: function(breed, name) {
    this.breed = breed
    this.name = name
  },
  call: function(name) {
    return this.name === name ? this.bark() : ''
  },
  toString: function() {
    return this.breed + ' ' + this.name
  }
})

// All arguments passed to the constructor are forwarded to the `initialize`
// method of instance.

var pet = new Pet('Labrador', 'Benzy')
pet.toString()          // 'Labrador Benzy'
pet.call('doggy')       // ''
pet.call('Benzy')       // 'Ruff! Ruff!'
```


### Object composition

```js
// In some programs recombining reusable pieces of code is a better option:

var HEX = Base.extend({
  hex: function hex() {
    return '#' + this.color
  }
})

var RGB = Base.extend({
  red: function red() {
    return parseInt(this.color.substr(0, 2), 16)
  },
  green: function green() {
    return parseInt(this.color.substr(2, 2), 16)
  },
  blue: function blue() {
    return parseInt(this.color.substr(4, 2), 16)
  }
})

var CMYK = Base.extend(RGB.prototype, {
  black: function black() {
    var color = Math.max(Math.max(this.red(), this.green()), this.blue())
    return (1 - color / 255).toFixed(4)
  },
  magenta: function magenta() {
    var K = this.black();
    return (((1 - this.green() / 255).toFixed(4) - K) / (1 - K)).toFixed(4)
  },
  yellow: function yellow() {
    var K = this.black();
    return (((1 - this.blue() / 255).toFixed(4) - K) / (1 - K)).toFixed(4)
  },
  cyan: function cyan() {
    var K = this.black();
    return (((1 - this.red() / 255).toFixed(4) - K) / (1 - K)).toFixed(4)
  }
})

// Composing `Color` prototype out of reusable components:
var Color = Base.extend(HEX.prototype, RGB.prototype, CMYK.prototype, {
  initialize: function initialize(color) {
    this.color = color
  }
})

var pink = Color.new('FFC0CB')
// RGB
pink.red()        // 255
pink.green()      // 192
pink.blue()       // 203

// CMYK
pink.magenta()    // 0.2471
pink.yellow()     // 0.2039
pink.cyan()       // 0.0000
```

### Combining composition & inheritance

```js
var Pixel = Color.extend({
  initialize: function initialize(x, y, color) {
    Color.initialize.call(this, color)
    this.x = x
    this.y = y
  },
  toString: function toString() {
    return this.x + ':' + this.y + '@' + this.hex()
  }
})

var pixel = new Pixel(11, 23, 'CC3399')
pixel.toString()              // 11:23@#CC3399
Pixel.prototype.isPrototypeOf(pixel)    // true

// Pixel instances inhertis from `Color`
Color.prototype.isPrototypeOf(pixel)    // true

// In fact `Pixel.prototype` itself inherits from `Color.prototype`, remember just simple and
// pure prototypal inheritance where objects inherit from objects.
Color.prototype.isPrototypeOf(Pixel.prototype)    // true
```

### TODO

This is a list of things I may introduce in newer versions.

* Add `testling` badge
* Add examples about private variables (closures)
* Add guidelines on how to use the lib for common cases
* Add `requirejs` test
* Better merge? in-depth merge (like lodash). Two different `extend` methods ?
* Do a full prototype chaining ?
```js
var Extra = Base.extend(Foo, Bar, {toto: true});
var instance = new Extra();
Base.prototype.isPrototypeOf(instance); // true :D
Extra.prototype.isPrototypeOf(instance); // true :D
Foo.prototype.isPrototypeOf(instance); // false :(
Bar.prototype.isPrototypeOf(instance); // false :(
```
* Clone inner objects:
```js
var someBase = {
  outer: {
    inner: 1
  }
}
var Extra = Base.extend(Foo, Bar, someBase);
var e = new Extra();
e.outer.inner = 3;
someBase.outer.inner === 1; // should be true
```


History.md
# Changes #

## 1.1.0 / 2015-06-21
  - Fixed #15: Added legacy `.new()` function
  - Fixed deprecated warning on package.json

## 1.0.2 / 2015-04-07
  - Fixed #16: instances properties defined through objects are not mutable

## 1.0.1 / 2015-04-03
  - Updated bower.json authors
  - I messed up publishing with npm so I need to create this version

## 1.0.0 / 2015-03-02

  - Mocha + should for testing
  - We now mess with the prototypes so we can call `new Foo()` instead of
      `Foo.new()`. This means that most of the calls that were made with `Foo`
      are now made with `Foo.prototype` except `extend` which is still made on
      the object and not on its prototype.
  - `instanceof` should now be working but this feature is not tested

## 0.3.3 / 2014-11-02

  - Updated requirejs compatibility

## 0.3.2 / 2011-12-29

  - Fix `package.json` so that npm no longer complains.

## 0.3.1 / 2011-12-29

  - Enable compatibility with latest nodes.
  - Enable travis tests.

## 0.3.0 / 2011-08-14

  - Introduce `megre` method.

## 0.2.2 / 2011-07-03 ##

  - Quote method `new` for pre ES5 JS engines so that library can be loaded and
    method can be aliased with `create` for example.

## 0.2.1 / 2011-06-28 ##

 - Removed obsolete package dependency.

## 0.2.0 / 2011-06-28 ##

 - Adding `initialize` method for that is called by `new` on each instance
   initialization.
 - Documentation!

## 0.1.0 / 2011-06-27 ##

 - Reducing API complexity by removing experimental `Base` as function form.
 - Change meaning of `.new` that returning new instance of the class instead
   of direct decedent of a target.
 - Change `extend` so that it may take multiple property maps.

## 0.0.3 / 2011-06-21 ##

 - Fixing delegation to `Base.new`.

## 0.0.2 / 2011-06-15 ##

  - Improve readme.
  - Bug fix.

## 0.0.1 / 2011-06-15 ##

  - Initial release


tests/browser.js
describe('#isPrototypeOf', function() {
  it('Base is a prototype of Base.new()', function() {
    expect(Base.isPrototypeOf(Base.new())).toBe(true);
  });
  it('Base is a prototype of Base.extend()', function() {
    expect(Base.isPrototypeOf(Base.extend())).toBe(true);
  });
  it('Base is a prototoype of Base.extend().new()', function() {
    expect(Base.isPrototypeOf(Base.extend().new())).toBe(true);
  });
  it('Base.extend() in not prototype of Base.extend()', function() {
    expect(Base.extend().isPrototypeOf(Base.extend())).toBe(false);
  });
  it('Base.extend() in not prototype of Base.extend().new()', function() {
    expect(Base.extend().isPrototypeOf(Base.extend().new())).toBe(false);
  });
  it('Base.new() is not prototype of Base.new()', function() {
    expect(Base.new().isPrototypeOf(Base.new())).toBe(false);
  });
  it('Base.new() is not prototype of Base.extend()', function() {
    expect(Base.new().isPrototypeOf(Base.extend())).toBe(false);
  });
  it('Base.extend() is not prototype of Base.new()', function() {
    expect(Base.extend().isPrototypeOf(Base.new())).toBe(false);
  });
});

describe('#inheritance', function() {
  var Parent, Child;
  beforeEach(function() {
    Parent = Base.extend({
      name: "parent",
      method: function() {
        return "hello " + this.name;
      }
    });
    Child = Parent.extend({
      name: "child"
    });
  });
  it('Parent name is parent', function() {
    expect(Parent.name).toBe("parent");
  });
  it('method works on prototype', function() {
    expect(Parent.method()).toBe("hello parent");
  });
  it('Parent instance inherits name', function() {
    expect(Parent.new().name).toBe(Parent.name);
  });
  it('method behaves same on the prototype', function() {
    expect(Parent.new().method()).toBe(Parent.method());
  });
  it('Parent decedent inherits name', function() {
    expect(Parent.extend({}).name).toBe(Parent.name);
  });
  it('Child overides name', function() {
    expect(Child.name).not.toBe(Parent.name);
  });
  it('Child intsances inherit name', function() {
    expect(Child.new().name).toBe(Child.name);
  });
  it('Child decedents inherit name', function() {
    expect(Child.extend().name).toBe(Child.name);
  });

  it('Child inherits method', function() {
    expect(Child.method).toBe(Parent.method);
  });
  it('Child decedent inherit method', function() {
    expect(Child.extend().method).toBe(Parent.method);
  });
  it('Child instances inherit method', function() {
    expect(Child.new().method).toBe(Parent.method);
  });

  it('method refers to instance proprety', function() {
    expect(Child.method()).toBe("hello child");
  });
  it('method may be overrided', function() {
    expect(Child.extend({
      name: "decedent"
    }).new().method()).toBe("hello decedent");
  });
});

describe('#prototype immutability', function() {
  it('Base prototype is imutable', function() {
    var tmp = Base.extend;
    var foo = function() {};
    Base.extend = foo;
    expect(Base.extend).toBe(tmp);
  });
  it('Base prototype is non-configurable', function() {
    var tmp = Base.foo;
    Base.foo = "bar";
    expect(Base.foo).toBe(tmp);
  });
  it('Can\'t delete properties on prototype', function() {
    var tmp = Base.new;
    delete Base.new;
    expect(Base.new).toBe(tmp);
  });
});

describe('#extended prototype immutability', function() {
  var Foo, Bar;
  beforeEach(function() {
    Foo = Base.extend({
      name: 'hello',
      rename: function rename(name) {
        this.name = name;
      },
      nothing: function nothing() {}
    });
    Bar = Foo.extend({
      rename: function rename() {
        return this.name;
      }
    });
  });
  it('Can\'t change prototype properties', function() {
    var tmp = Foo.extend;
    Foo.extend = function() {}
    expect(Foo.extend).toBe(tmp);
  });

  it('Can\'t add prototype properties', function() {
    var tmp = Foo.foo;
    Foo.foo = 'bar';
    expect(Foo.foo).toBe(tmp);
  });

  it('Can\'t remove prototype properties', function() {
    var tmp = Foo.name;
    delete Foo.name;
    expect(Foo.name).toBe(tmp);
  });

  it("Method's can't mutate prototypes", function() {
    var tmp = Foo.name;
    Foo.rename("new name");
    expect(Foo.name).toBe(tmp);
    expect(Foo.name).not.toBe("new name");
  });

  it("properties may be overided on decedents", function() {
    expect(Bar.rename()).toBe(Foo.name);
    expect(Bar.rename).not.toBe(Foo.rename);
    expect(Bar.nothing).toBe(Foo.nothing);
  });
});

describe('#instance mutability', function() {
  var Foo, f1;
  beforeEach(function() {
    Foo = Base.extend({
      name: "foo",
      init: function init(number) {
        this.number = number;
      }
    });
    f1 = Foo.new();
  });
  it('instance is mutable', function() {
    f1.alias = 'f1';
    expect(f1.alias).toBe('f1');
  });
  it('own properties are deletable', function() {
    f1.alias = 'f1';
    delete f1.alias;
    expect('alias' in f1).toBe(false);
  });
  it('methods can mutate instance\'s own properties', function() {
    f1.init(1);
    expect(f1.number).toBe(1);
  });
});

describe('#super', function() {
  var Foo, Bar, bar;
  beforeEach(function() {
    Foo = Base.extend({
      initialize: function Foo(options) {
        this.name = options.name;
      }
    });

    Bar = Foo.extend({
      initialize: function Bar(options) {
        Foo.initialize.call(this, options);
        this.type = 'bar';
      }
    });

    bar = Bar.new({
      name: 'test'
    });
  });

  it('Bar is prototype of Bar.new', function() {
    expect(Bar.isPrototypeOf(bar)).toBe(true);
  });
  it('Foo is prototype of Bar.new', function() {
    expect(Foo.isPrototypeOf(bar)).toBe(true);
  });
  it('Base is prototype of Bar.new', function() {
    expect(Base.isPrototypeOf(bar)).toBe(true);
  });
  it('bar initializer was called', function() {
    expect(bar.type).toBe('bar');
  });
  it('bar initializer called Foo initializer', function() {
    expect(bar.name).toBe('test');
  });
});


test/test-selfish.js
(function() {
  "use strict";

  var selfish = require("../selfish");
  var should = require("should");
  var Base = selfish.Base;

  describe("#isPrototypeOf", function() {
    it("Base is a prototype of new Base()", function() {
      Base.prototype.isPrototypeOf(new Base()).should.be.true;
    });
    it("Base is a prototype of Base.extend()", function() {
      Base.prototype.isPrototypeOf(Base.extend().prototype).should.be.true;
    });
    it("Base is a prototype of new (Base.extend())", function() {
      Base.prototype.isPrototypeOf(new(Base.extend())()).should.be.true;
    });
    it("Base.extend() in not prototype of Base.extend()", function() {
      Base.extend().prototype.isPrototypeOf(Base.extend().prototype).should.be.false;
    });
    it("Base.extend() in not prototype of new (Base.extend())", function() {
      Base.extend().prototype.isPrototypeOf(new(Base.extend())()).should.be.false;
    });
    it("new Base() is not prototype of new Base()", function() {
      (new Base()).isPrototypeOf(new Base()).should.be.false;
    });
    it("new Base() is not prototype of Base.extend()", function() {
      (new Base()).isPrototypeOf(Base.extend().prototype).should.be.false;
    });
    it("Base.extend() is not prototype of new Base()", function() {
      Base.extend().prototype.isPrototypeOf(new Base()).should.be.false;
    });
  });

  describe("#Inheritance", function() {
    var Parent, Child;
    before(function() {
      Parent = Base.extend({
        name: "parent",
        method: function() {
          return "hello " + this.name;
        }
      });
      Child = Parent.extend({
        name: "child"
      });
    });

    describe("Parent prototype", function() {
      it("Parent.prototype name is parent", function() {
        Parent.prototype.name.should.be.eql("parent");
      });
      it("method works on prototype", function() {
        Parent.prototype.method().should.be.eql("hello parent");
      });
    });

    describe("Parent instances", function() {
      var p;
      before(function() {
        p = new Parent();
      });
      it("Parent instance inherits name", function() {
        var p = new Parent();
        p.name.should.be.eql(Parent.prototype.name);
      });
      it("method behaves same on the prototype", function() {
        p.method().should.be.eql(Parent.prototype.method());
      });
    });

    it("Parent decedent inherits name", function() {
      Parent.extend({}).prototype.name.should.be.eql(Parent.prototype.name);
    });

    describe("Child prototype", function() {
      it("Child overrides name", function() {
        Child.prototype.name.should.not.be.eql(Parent.prototype.name);
      });
      it("method refers to instance property", function() {
        Child.prototype.method().should.be.eql("hello child");
      });
      it("Child inherits method", function() {
        Child.prototype.method.should.be.eql(Parent.prototype.method);
      });
    });

    describe("Child instances", function() {
      var c;
      before(function() {
        c = new Child();
      });

      it("Child instances inherit name", function() {
        c.name.should.be.eql(Child.prototype.name);
      });
      it("Child instances inherit method", function() {
        c.method.should.be.eql(Parent.prototype.method);
      });
    });

    it("Child decedents inherit name", function() {
      Child.extend().prototype.name.should.be.eql(Child.prototype.name);
    });

    it("Child decedent inherit method", function() {
      Child.extend().prototype.method.should.be.eql(Parent.prototype.method);
    });

    it("method may be overridden", function() {
      var CC = Child.extend({
        name: "decedent"
      });
      var cc = new CC();
      cc.method().should.be.eql("hello decedent");
    });
  });

  describe("#prototype immutability", function() {
    it("Base is immutable", function() {
      (function() {
        Base.extend = null;
      }).should.throw();
    });
    it("Base prototype is immutable", function() {
      (function() {
        Base.prototype.foo = null;
      }).should.throw();
    });
    it("Base prototype is non-configurable", function() {
      (function() {
        Base.foo = null;
      }).should.throw();
    });
    it("Can't delete properties on Base", function() {
      (function() {
        delete Base.extend;
      }).should.throw();
    });
  });

  describe("#extended prototype immutability", function() {
    var Foo, Bar;
    before(function() {
      Foo = Base.extend({
        name: "hello",
        rename: function rename(name) {
          this.name = name;
        },
        nothing: function nothing() {}
      });
      Bar = Foo.extend({
        rename: function rename() {
          return this.name;
        }
      });
    });
    it("Can't change prototype properties", function() {
      (function() {
        Foo.extend = null;
      }).should.throw();
    });

    it("Can't add properties", function() {
      (function() {
        Foo.prototype.bar = null;
      }).should.throw();
      (function() {
        Foo.bar = null;
      }).should.throw();
    });

    it("Can't remove prototype properties", function() {
      (function() {
        delete Foo.prototype.name;
      }).should.throw();
    });

    it("Method's can't mutate prototypes", function() {
      (function() {
        Foo.prototype.rename("new name");
      }).should.throw();
    });

    it("properties may be overridden on decedents", function() {
      Bar.prototype.rename().should.be.eql(Foo.prototype.name);
      Bar.prototype.rename.should.not.be.eql(Foo.prototype.rename);
    });
  });

  describe("#instance mutability", function() {
    var Foo, f1;
    before(function() {
      Foo = Base.extend({
        name: "foo",
        init: function init(number) {
          this.number = number;
        }
      });
      f1 = new Foo();
    });
    it("instance is mutable", function() {
      should(f1.alias).be.ko;
      f1.alias = "f1";
      f1.alias.should.be.eql("f1");
    });
    it("own properties are deletable", function() {
      f1.alias = "f1";
      ("alias" in f1).should.be.true;
      delete f1.alias;
      ("alias" in f1).should.be.false;
    });
    it("methods can mutate instance's own properties", function() {
      f1.init(1);
      f1.number.should.be.eql(1);
      f1.init(3);
      f1.number.should.be.eql(3);
    });
    it("properties can be directly mutated", function() {
      f1.name.should.be.eql("foo");
      (function() {
        f1.name = "lol";
      }).should.not.throw();
      f1.name.should.be.eql("lol");
    });
    describe("#Deeper inheritance mutability", function() {
      var A, B, C;
      before(function() {
        A = Base.extend({
          name: "A",
          hello: function() {
            return this.name;
          }
        });
        B = A.extend({
          varB: "B"
        });
        C = B.extend({
          varC: "C"
        });
      });
      it("A instance can modify name", function() {
        var a = new A();
        a.name.should.be.eql("A");
        a.hello().should.be.eql("A");
        (function() {
          a.name = "Foo";
        }).should.not.throw();
        a.name.should.be.eql("Foo");
        a.hello().should.be.eql("Foo");
      });
      it("B instance inherits A.name", function() {
        var b = new B();
        b.name.should.be.eql("A");
        b.hello().should.be.eql("A");
      });
      it("B instance varB is mutable", function() {
        var b = new B();
        b.varB.should.be.eql("B");
        (function() {
          b.varB = "Foo";
        }).should.not.throw();
        b.varB.should.be.eql("Foo");
      });
      it("B instance can modify name", function() {
        var c = new C();
        (function() {
          c.name = "Foo";
        }).should.not.throw();
        c.name.should.be.eql("Foo");
        c.hello().should.be.eql("Foo");
      });
      it("C instance varC is mutable", function() {
        var c = new C();
        c.varC.should.be.eql("C");
        (function() {
          c.varC = "Foo";
        }).should.not.throw();
        c.varC.should.be.eql("Foo");
      });
      it("C instance can modify name", function() {
        var c = new C();
        (function() {
          c.name = "Foo";
        }).should.not.throw();
        c.name.should.be.eql("Foo");
        c.hello().should.be.eql("Foo");
      });
      it("C instance can modify varB", function() {
        var c = new C();
        (function() {
          c.varB = "Foo";
        }).should.not.throw();
        c.varB.should.be.eql("Foo");
      });
    });
  });

  describe("#Super", function() {
    var Foo, Bar, bar, C, c;
    before(function() {
      Foo = Base.extend({
        initialize: function Foo(options) {
          this.name = options.name;
        }
      });

      Bar = Foo.extend({
        initialize: function Bar(options) {
          Foo.prototype.initialize.call(this, options);
          this.type = "bar";
        }
      });

      C = Foo.extend({
        initialize: function C(options) {
          Foo.prototype.initialize.call(this, options);
          this.type = "C";
        }
      });

      bar = new Bar({
        name: "test"
      });

      c = new C({
        name: "not bar"
      });
    });

    it("Bar is prototype of new Bar()", function() {
      Bar.prototype.isPrototypeOf(bar).should.be.true;
    });
    it("Foo is prototype of new Bar()", function() {
      Foo.prototype.isPrototypeOf(bar).should.be.true;
    });
    it("Base is prototype of new Bar()", function() {
      Base.prototype.isPrototypeOf(bar).should.be.true;
    });
    it("C is not prototype of new Bar()", function() {
      C.prototype.isPrototypeOf(bar).should.be.false;
    });
    it("bar initializer was called", function() {
      bar.type.should.be.eql("bar");
    });
    it("bar initializer called Foo initializer", function() {
      bar.name.should.be.eql("test");
    });
  });

  describe("#Complex inheritance", function() {
    var HEX, RGB, CMYK, Color;
    before(function() {
      HEX = Base.extend({
        hex: function hex() {
          return "#" + this.color;
        }
      });

      RGB = Base.extend({
        red: function red() {
          return parseInt(this.color.substr(0, 2), 16);
        },
        green: function green() {
          return parseInt(this.color.substr(2, 2), 16);
        },
        blue: function blue() {
          return parseInt(this.color.substr(4, 2), 16);
        }
      });

      CMYK = Base.extend(RGB.prototype, {
        black: function black() {
          var color = Math.max(Math.max(this.red(), this.green()), this.blue());
          return (1 - color / 255).toFixed(4);
        },
        cyan: function cyan() {
          var K = this.black();
          return (((1 - this.red() / 255).toFixed(4) - K) / (1 - K)).toFixed(4);
        },
        magenta: function magenta() {
          var K = this.black();
          return (((1 - this.green() / 255).toFixed(4) - K) / (1 - K)).toFixed(4);
        },
        yellow: function yellow() {
          var K = this.black();
          return (((1 - this.blue() / 255).toFixed(4) - K) / (1 - K)).toFixed(4);
        }
      });

      Color = Base.extend(HEX.prototype, RGB.prototype, CMYK.prototype, {
        initialize: function Color(hex) {
          this.color = hex;
        }
      });
    });

    it("CMYK includes RGB but RGB is not his prototype", function() {
      RGB.prototype.isPrototypeOf(CMYK.prototype).should.be.false;
      var c = new CMYK();
      RGB.prototype.isPrototypeOf(c).should.be.false;
      CMYK.prototype.isPrototypeOf(c).should.be.true;
    });

    it("Color have hex, red and yellow", function() {
      should(Color.prototype.hex).be.a.Function;
      should(Color.prototype.red).be.a.Function;
      should(Color.prototype.yellow).be.a.Function;
    });

    it("Color instance have hex, red and yellow", function() {
      var c = new Color("#ff0000");
      should(c.hex).be.a.Function;
      should(c.red).be.a.Function;
      should(c.yellow).be.a.Function;
    });
  });
})();


saucelabs.conf.js
module.exports = function(config) {
  var customLaunchers = {
    sl_chrome: {
      base: 'SauceLabs',
      browserName: 'chrome',
      platform: 'Windows 7',
      version: '40'
    },
    sl_firefox: {
      base: 'SauceLabs',
      browserName: 'firefox',
      version: '35'
    },
    sl_opera: {
      base: 'SauceLabs',
      browserName: 'opera',
      version: '11'
    },
    sl_ios_safari: {
      base: 'SauceLabs',
      browserName: 'iphone',
      platform: 'OS X 10.10',
      version: '8.0'
    },
    sl_osx_safari: {
      base: 'SauceLabs',
      browserName: 'safari',
      platform: 'OS X 10.9',
      version: '7'
    },
    sl_ie_9: {
      base: 'SauceLabs',
      browserName: 'internet explorer',
      platform: 'Windows 7',
      version: '9'
    },
    sl_ie_10: {
      base: 'SauceLabs',
      browserName: 'internet explorer',
      platform: 'Windows 8',
      version: '10'
    },
    sl_ie_11: {
      base: 'SauceLabs',
      browserName: 'internet explorer',
      platform: 'Windows 8.1',
      version: '11'
    }
  };

  config.set({
    basePath: '',
    frameworks: ['jasmine'],
    files: [
      'selfish.js',
      'tests/browser.js'
    ],
    exclude: [],
    preprocessors: {},
    reporters: ['dots', 'sauceLabs'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    sauceLabs: {
      testName: 'Web App Unit Tests',
      recordScreenshots: false,
      connectOptions: {
        port: 5757,
        logfile: 'sauce_connect.log'
      }
    },
    // Increase timeout in case connection in CI is slow
    captureTimeout: 180000,
    customLaunchers: customLaunchers,
    browsers: Object.keys(customLaunchers),
    // Continuous Integration mode
    singleRun: true
  });
};


selfish.js
/* vim:set ts=2 sw=2 sts=2 expandtab */
/*jshint undef: true es5: true node: true devel: true evil: true
         forin: true latedef: false supernew: true */
/*global define: true */

!(typeof define !== "function" ? function(_, $) {
  $(null, typeof exports !== "undefined" ? exports : window);
} : define)("selfish", function(require, exports) {

  "use strict";

  /**
   * Merges all the properties of the passed objects into `this` instance (This
   * method can be used on instances only as prototype objects are frozen).
   *
   * If two or more argument objects have own properties with the same name,
   * the property is overridden, with precedence from right to left, implying,
   * that properties of the object on the left are overridden by a same named
   * property of the object on the right.
   *
   * @examples
   *
   *    var Pet = Dog.extend({
   *      initialize: function initialize(options) {
   *        this.name = options.name;
   *      },
   *      call: function(name) {
   *        return this.name === name ? this.bark() : '';
   *      },
   *      name: null
   *    });
   *    var pet = new Pet({ name: 'Benzy', breed: 'Labrador' });
   *    pet.call('Benzy');   // 'Ruff! Ruff!'
   */
  var merge = function selfishMerge() {
    var descriptor = {}, property;
    var that = Array.prototype.shift.call(arguments) || {};
    that.descriptors = {};
    Array.prototype.forEach.call(arguments, function(properties) {
      Object.getOwnPropertyNames(properties).forEach(function(name) {
        property = Object.getOwnPropertyDescriptor(properties, name);
        // This happens to be a problem for the constructor variable
        // A proper solution has to be found. This is only a fix
        if (property.enumerable) {
          descriptor[name] = property;
          descriptor[name].configurable = true;
          descriptor[name].enumerable = true;
          descriptor[name].writable = true;
          that.descriptors[name] = descriptor[name];
        }
      });
    });
    Object.defineProperties(this, descriptor);
    return this;
  };

  exports.merge = merge;

  /**
   * This is the main constructor. It should be used by any other object.
   * Always calling the `initialize` function makes it easier to keep
   * a coherent inheritance tree.
   */
  var Base = function BaseConstructor() {
    this.initialize.apply(this, arguments);
  };

  /**
   * When new instance of the this prototype is created its `initialize`
   * method is called with all the arguments passed to the `new`. You can
   * override `initialize` to set up an instance.
   */
  Base.prototype.initialize = function BaseInitialize() {};

  /**
   * Constructs an instance. An alternative maker function to using new keyword.
   *
   * @examples
   *
   *    var Pet = Dog.extend({
   *      initialize: function initialize(options) {
   *        this.name = options.name;
   *      },
   *      call: function(name) {
   *        return this.name === name ? this.bark() : '';
   *      },
   *      name: null
   *    });
   *    var pet = Pet.new({ name: 'Benzy', breed: 'Labrador' });
   *    pet.call('Benzy');   // 'Ruff! Ruff!'
   */
  Base.new = function() {
    var instance = Object.create(Base.prototype);
    Base.apply(instance, arguments)
    return instance;
  };

  /**
   * Takes any number of argument objects and returns frozen, composite object
   * that inherits from `this` object and combines all of the own properties of
   * the argument objects. (Objects returned by this function are frozen as
   * they are intended to be used as types).
   *
   * If two or more argument objects have own properties with the same name,
   * the property is overridden, with precedence from right to left, implying,
   * that properties of the object on the left are overridden by a same named
   * property of the object on the right.
   *
   * When extending selfish prototypes remember to call it on their `prototype`
   * instead of the type itself.
   *
   * This is the only function that should not be placed in the prototype as it
   * is relative to the object.
   * @examples
   *
   *    // ## Object composition ##
   *
   *    var HEX = Base.extend({
   *      hex: function hex() {
   *        return '#' + this.color;
   *      }
   *    });
   *
   *    var RGB = Base.extend({
   *      red: function red() {
   *        return parseInt(this.color.substr(0, 2), 16);
   *      },
   *      green: function green() {
   *        return parseInt(this.color.substr(2, 2), 16);
   *      },
   *      blue: function blue() {
   *        return parseInt(this.color.substr(4, 2), 16);
   *      }
   *    });
   *
   *    var CMYK = Base.extend(RGB.prototype, {
   *      black: function black() {
   *        var color = Math.max(Math.max(this.red(), this.green()), this.blue());
   *        return (1 - color / 255).toFixed(4);
   *      },
   *      cyan: function cyan() {
   *        var K = this.black();
   *        return (((1 - this.red() / 255).toFixed(4) - K) / (1 - K)).toFixed(4);
   *      },
   *      magenta: function magenta() {
   *        var K = this.black();
   *        return (((1 - this.green() / 255).toFixed(4) - K) / (1 - K)).toFixed(4);
   *      },
   *      yellow: function yellow() {
   *        var K = this.black();
   *        return (((1 - this.blue() / 255).toFixed(4) - K) / (1 - K)).toFixed(4);
   *      }
   *    });
   *
   *    var Color = Base.extend(HEX.prototype, RGB.prototype, CMYK.prototype, {
   *      initialize: function Color(color) {
   *        this.color = color;
   *      }
   *    });
   *
   *    // ## Prototypal inheritance ##
   *
   *    var Pixel = Color.extend({
   *      initialize: function Pixel(x, y, hex) {
   *        Color.initialize.call(this, hex);
   *        this.x = x;
   *        this.y = y;
   *      },
   *      toString: function toString() {
   *        return this.x + ':' + this.y + '@' + this.hex();
   *      }
   *    });
   *
   *    var pixel = new Pixel(11, 23, 'CC3399');
   *    pixel.toString(); // 11:23@#CC3399
   *
   *    pixel.red();      // 204
   *    pixel.green();    // 51
   *    pixel.blue();     // 153
   *
   *    pixel.cyan();     // 0.0000
   *    pixel.magenta();  // 0.7500
   *    pixel.yellow();   // 0.2500
   *
   */
  Base.extend = function selfishExtend() {
    var dependencies = [];
    /*
     * It is actually important to redefine a constructor here. Even if
     * the body is always the same
     */
    var constructor = function() { // Call initialize by default if possible
      var name;
      if (constructor.descriptors) {
        for (name in constructor.descriptors) {
          if (constructor.descriptors.hasOwnProperty(name)) {
            Object.defineProperty(this, name, constructor.descriptors[name]);
          }
        }
      }
      this.initialize.apply(this, arguments);
    };
    Array.prototype.forEach.call(arguments, function(dependency) {
      dependencies.push(dependency);
    });
    // Copy the prototype methods that allows us to do inheritance
    constructor.extend = this.extend;
    constructor.new = function() {
      var instance = Object.create(constructor.prototype);
      constructor.apply(instance, arguments)
      return instance;
    };

    // Generate the prototype and extend it
    var descriptor = Object.create(this.prototype);
    // We want to keep the prototype properties too
    dependencies.unshift(this.prototype);
    dependencies.unshift(constructor);
    merge.apply(descriptor, dependencies);

    // Freeze the new created prototype
    constructor.prototype = Object.freeze(descriptor);
    return Object.freeze(constructor);
  };

  // Freeze the prototype as well as the constructor
  Base.prototype = Object.freeze(Base.prototype);
  exports.Base = Object.freeze(Base);

});
